TaboolaSDKExampleSwift/TaboolaSDKExampleSwift/SDK Native/NativeViewController.swift

```diff
@@ -10,60 +10,110 @@ import Foundation
+                guard let newItems = response.items else { return }
+                self.collectionView.performBatchUpdates({
+                    guard let resultsSize = self.itemsArray?.count else { return }
+                    self.itemsArray?.add(newItems)
+                    let arrayWithIndexPaths = NSMutableArray()
+                    for i in 0...(resultsSize + newItems.count) {
+                        arrayWithIndexPaths.add(NSIndexPath(row: i, section: 0))
+                    }
+                    self.collectionView.insertItems(at: arrayWithIndexPaths as! [IndexPath])
+                }, completion: nil)
+            }
+        }, onFailure: { (error) in
+            
+        })
     }
 }
 
-extension NativeViewController: UITableViewDataSource {
-    func numberOfSections(in tableView: UITableView) -> Int {
+extension NativeViewController: UICollectionViewDelegate, UICollectionViewDataSource {
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
         return Constants.totalSectionsNative
     }
     
-    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return Constants.totalRowsNative
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let itemsArray = itemsArray else { return 0 }
+        return itemsArray.count
     }
     
-    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        let cell = tableView.dequeueReusableCell(withIdentifier: "TextCell") ?? UITableViewCell(style: .default, reuseIdentifier: "TextCell")
-        cell.textLabel?.text = Constants.nativeText
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "NativeCollectionViewCell", for: indexPath)
+        if let cell = cell as? NativeViewControllerCell {
+            cell.attributionButton.addTarget(self, action: #selector(clickedTaboolaAttribution), for: .touchDown)
+            guard let itemsArray = itemsArray else { return cell }
+            if let item = itemsArray[indexPath.row] as? TBLNativeItem {
+                if let imageUrl = item.extraDataDictionary()["imageUrl"] as? NSString {
+                    DispatchQueue.global(qos: .background).async {
+                        guard let data = NSData.init(contentsOf: NSURL(string: imageUrl as String)! as URL) else { return }
+                        DispatchQueue.main.async {
+                            cell.imageView.image = UIImage(data: data as Data)
+                        }
+                    }
+                    cell.imageView.isHidden = false
+                    cell.tbImageView.isHidden = true
+                    cell.tbImageView.isUserInteractionEnabled = false
+                    cell.imageView.isUserInteractionEnabled = true
```

Comment:
Another point to think about (*outside the scope of this PR): tbImageView and imageView seem like 2 frames that you use for 2 different cases. Why not use just 1 frame for both the real image if succesfully retrieved and the fallback image if there was a problem.

Reply:
won't happen now. needs to be discussed .